### ISSUE TITLE
Experiment end date can be incorrect if overridden in external config

### ISSUE THREAD

**Issue by scholtzan** [mozilla/experimenter#6200]
External configuration files allow to change the `end_date` of the experiment. For example, for `1-click-pin-experiment` see https://github.com/mozilla/jetstream-config/blob/cbf385c231ce8d830ccf22153a753c580047ff2e/1-click-pin-experiment.toml#L6

Experimenter still shows the end date that was configured when the experiment was initially created:

![Screen Shot 2021-08-12 at 14 44 12](https://user-images.githubusercontent.com/1239705/129274178-683fa511-492a-487f-8f57-8791b527c1a2.png)

I've had a couple of cases where stakeholders/people looking at the result were confused and thought results were missing or experiments should have already completed.

This is related to an issue we found with configs overriding the `reference_branch`: https://github.com/mozilla/experimenter/issues/6129
One way of solving this is adding the `end_date` to the experiment metadata that Jetstream produces and use it as the source of truth.

**Comment by LZoog** [comment]
@scholtzan What else can be overridden in the config?

We show the end date on the summary page too. Right now the metadata is returned with the analysis fetch, and if the experiment has been launched, we hit that endpoint to check if analysis data is available yet; so while it'd require some refactoring we _can_ rework the end date to point at the metadata on the analysis page and summary page, but is this something that will keep coming up with other experiment properties? If it is, I wonder if we should or if it's plausible to pull in that metadata and change the experiment data from within Experimenter so it matches and then we don't have to have a bunch of conditionals around if there's overrides in the metadata. 😕

**Comment by scholtzan** [comment]
The config can override `reference_branch`, `end_date`, `start_date`, and `proposed_enrollment`. I think more properties could be added to that list in the future, but nothing is planned at the moment.
Yeah, updating the experiment settings in Experimenter would probably be ideal.

**Comment by lmorchard** [comment]
This is probably a function of me not yet being familiar with this part of Nimbus, but: Does this point to a general problem with source of truth for an experiment?

*Should* the end date of an experiment be overridden in external config? What's the use case & rationale for that? 

Not asserting that this *shouldn't* happen, so much as trying to understand when & why it would happen. e.g. Is it expected that Experimenter Console should poll external configs to update its database and published DTOs? Do changes need to be pushed to Remote Settings based on external configs?

**Comment by scholtzan** [comment]
One of the main use cases for why this was made possible is that sometimes stakeholders or DS want to change the end date or reference branch after the experiment has already completed. Changing the configuration via [jetstream-config](https://github.com/mozilla/jetstream-config) will ensure that the analysis for the experiment gets rerun with the new configuration parameters when the updates get merged.

If the changes are only done via Experimenter, then Jetstream will either not be aware of any changes (if the experiment has already ended) or it will not realize that changes have been made that affect previous analysis runs.

**Comment by LZoog** [comment]
In that case, the experiment "end date" in the header is technically correct then, but the analysis data is for a different date set?

Maybe this isn't as big of an issue as I thought it might be then - we could potentially have a different section at the top of the Analysis page with information about date periods for the analysis shown, if I'm understanding correctly. But, how/why can `proposed_enrollment` be overridden and how does that affect analysis data?

**Comment by lmorchard** [comment]
> One of the main use cases for why this was made possible is that sometimes stakeholders or DS want to change the end date or reference branch after the experiment has already completed. Changing the configuration via jetstream-config will ensure that the analysis for the experiment gets rerun with the new configuration parameters when the updates get merged.

More thinking out loud, sorry if this is obvious stuff:

To me, this sounds like we "pass the ball" from Experimenter Console to Jetstream after experiment launch. (i.e. the source of truth has moved, no more edits allowed from Experimenter Console)

So, if these properties are exposed via the analysis endpoint, then maybe we should override display of the local database values with the latest from analysis? Maybe even update them locally?

(Adding a to-do item to finally get around to learning exactly how all the Jetstream stuff works 😅)

**Comment by scholtzan** [comment]
> how/why can proposed_enrollment be overridden and how does that affect analysis data?

Looking through existing configs, this hasn't actually been used so far. I guess it only makes sense if for some reason the enrollment period needs to be decreased. Generally, jetstream only considers clients as enrolled in an experiment that have sent an `enroll` event some time between the `start_date` and `start_date` + `proposed_enrollment` (`proposed_enrollment` = enrollment period in days).

**Comment by scholtzan** [comment]
> To me, this sounds like we "pass the ball" from Experimenter Console to Jetstream after experiment launch. (i.e. the source of truth has moved, no more edits allowed from Experimenter Console)

Yes, that sounds right 👍 

> So, if these properties are exposed via the analysis endpoint, then maybe we should override display of the local database values with the latest from analysis? Maybe even update them in the DB locally?

I think that would be ideal. Not sure how much work that would be on the Experimenter side. It might also possibly be good enough to just show the parameters the analysis is based on in some section on the Results page.

**Comment by LZoog** [comment]
> So, if these properties are exposed via the analysis endpoint, then maybe we should override display of the local database values with the latest from analysis? Maybe even update them in the DB locally?

Yeah, that's what I was thinking with my "I wonder if we should or if it's plausible to pull in that metadata and change the experiment data from within Experimenter" comment. But... if the changes _only_ impact the analysis data, updating the DB might be more trouble than it's worth IMO, if we can simply display the overrides in a nice way at the top of the Results page, like with "Analysis date range: X - X" and probably also mention if there was a control branch/proposed enrollment override when one exists. 

Probably should get @jaredlockhart's opinion.

**Comment by data-sync-user** [comment]
➤ Jared Lockhart commented:

Blocked pending ruther discussion with Lauren Zugai Anna Scholtz Marina Samuel

**Comment by LZoog** [comment]
For posterity: we had a meeting today and basically said since this is scoped to the Analysis data, we're going to use the overrides if they exist in the Jetstram config and then display them at the top of the page in a Boostrap Alert just to make it obvious. We won't be updating the DB.

**Comment by scholtzan** [comment]
https://github.com/mozilla/jetstream/pull/860 adds an `external_config` JSON object to the metadata jetstream exports for each experiment. `external_config` is `null` if no external config exists or if the external config does not override any of the parameters:
![Screen Shot 2021-08-18 at 11 16 42](https://user-images.githubusercontent.com/1239705/129951895-02d26468-2dda-450e-9551-b5f77fb6964b.png)

The experiment parameters that have not been changed and are the same as in Experimenter will be `null`. I also added `skip` which indicates that analysis is skipped for an experiment and `url` which links to the config in Github.
Does this look reasonable?

**Comment by LZoog** [comment]
Thanks for your quick work @scholtzan, that looks perfect! 😄

> I also added `skip` which indicates that analysis is skipped for an experiment

If analysis is skipped, what's that mean in terms of the analysis data, will `overall` just come back as `null`? Or should we always check it for truthiness, such as `external_config.reference_branch && skip` rather than just check `external_config.reference_branch` if we wanted to pull out this referene branch?

**Comment by scholtzan** [comment]
If an analysis is skipped, then there won't be any analysis result data. So GCS won't have any JSON files for overall, weekly, daily, ... results. This is the case for the following experiment which has `skip=true`: https://experimenter.services.mozilla.com/nimbus/set-default-as-first-screen-100-roll-out

**Comment by LZoog** [comment]
Ah interesting, we could potentially use that to display "Experiment analysis was skipped" in the sidebar instead of "Experiment results not yet ready". Could that be helpful at all?
